antsibull/docsite_docs.py

```diff
@@ -0,0 +1,234 @@
+
+    :arg content: Content of a RST document.
+    :arg collection_name: Dotted collection name.
+    :returns: A tuple consisting of a list of RST labels and a list of error messages
+              (with line and column numbers).
+    '''
+    labels: t.Set[str] = set()
+    errors: t.List[t.Tuple[int, int, str]] = []
+    label_prefix = get_label_prefix(collection_name)
+    # Check label definitions
+    for row, line in enumerate(content.splitlines()):
+        m = _RST_LABEL_DEFINITION.match(line)
+        if m:
+            label = m.group(1)
+            if not label.startswith(label_prefix):
+                errors.append((
+                    row + 1,
+                    0,
+                    f'Label "{label}" does not start with expected prefix "{label_prefix}"'))
+            else:
+                labels.add(label)
+    return sorted(labels), errors
+
+
+def load_toctree(yaml_section: t.Dict[str, t.Any], section_index: int = 0
+                 ) -> t.Tuple[t.List[str], t.List[str]]:
+    errors: t.List[str] = []
+    toctree: t.List[str] = []
+    if 'toctree' in yaml_section:
+        for toctree_index, toctree_name in enumerate(yaml_section['toctree']):
+            if not isinstance(toctree_name, str):
+                errors.append(
+                    f'Toctree entry #{toctree_index} in section #{section_index}'
+                    f' is not a string')
+                continue
+            toctree.append(toctree_name)
+    return toctree, errors
+
+
+def load_section(yaml_section: t.Dict[str, t.Any], section_index: int = 0
+                 ) -> t.Tuple[t.Optional[Section], t.List[str]]:
+    errors: t.List[str] = []
+    missing = False
+    for required_key in ('title', ):
+        if required_key not in yaml_section:
+            errors.append(
+                f'Section #{section_index} has no "{required_key}" entry')
+            missing = True
+    if missing:
+        return None, errors
+    toctree, toctree_errors = load_toctree(yaml_section, section_index)
+    errors.extend(toctree_errors)
+    if not toctree:
+        errors.append(
+            f'Section #{section_index} has no content')
+        return None, errors
+    return Section(yaml_section['title'], toctree), errors
+
+
+def load_docsite_index(index_path: str) -> t.Tuple[t.List[Section], t.List[str]]:
+    '''Load a docsite extra-docs.yml file.
+
+    :arg index_path: Path to extra-docs.yml (does not need to exist).
+    :arg report_not_existing: Whether to report when extra-docs.yml does not exist.
+    :returns: A tuple consisting of a list of sections and a list of error messages.
+    :raises: DocsiteIndexError if extra-docs.yml does not exist
+    '''
+    sections: t.List[Section] = []
+    errors: t.List[str] = []
+
+    if not os.path.isfile(index_path):
+        raise DocsiteIndexError('extra-docs.yml does not exist')
+
+    try:
+        index = load_yaml_file(index_path)
```

Comment:
I note that this is synchronous code rather than async code.  The majority of time is going to be spent reading and writing the actual page data rather than the extra-docs.yml file so it's not the end of the world but it is a point where we could be doing things in parallel.  I assume you coded it as synchronous for easy reuse with antsibull-lint?

Reply:
The main reason is that the IO part of this function is done by `load_yaml_file`, which is not async either. It is also used to load routing data for collections, so eventually it should be converted to async, but once that's done adjusting the calling code is also not so hard, so I would defer adjusting this function until then. Is that ok for you?